nucleus/security/core/src/main/java/com/sun/enterprise/security/auth/login/ClientPasswordLoginModule.java

```diff
@@ -133,155 +136,160 @@ public void initialize(Subject subject, CallbackHandler callbackHandler, Map sha
      * @exception FailedLoginException if the authentication fails.
      * <p>
      *
-     * @exception LoginException if this <code>LoginModule</code> is unable to perform the authentication.
+     * @exception LoginException if this <code>LoginModule</code> is unable to perform the
+     * "authentication".
      */
+    @Override
     public boolean login() throws LoginException {
+        callbackHandlerNonNull();
 
-        // prompt for a username and password
-        if (callbackHandler == null) {
-            String failure = localStrings.getLocalString("login.nocallback",
-                    "Error: no CallbackHandler available to garner authentication information from the user");
-            throw new LoginException(failure);
-        }
+        // Try to get the username and password from the exchange mechanism, which is either TLS or an global variable
+        String incomingUsername = UsernamePasswordStore.getUsername();
+        char[] incomingPassword = UsernamePasswordStore.getPassword();
 
-        // Get the username from the exchange mechanism
-        String uname = UsernamePasswordStore.getUsername();
-        char[] pswd = UsernamePasswordStore.getPassword();
-        boolean doSet = false;
+        // Try to get the username and password from system properties if not provided by UsernamePasswordStore
 
-        // bugfix# 6412539
-        if (uname == null) {
-            uname = System.getProperty(LOGIN_NAME);
+        boolean doSet = false;
+        if (incomingUsername == null) {
+            incomingUsername = System.getProperty(LOGIN_NAME);
             doSet = true;
         }
-        if (pswd == null) {
+        if (incomingPassword == null) {
             if (System.getProperty(LOGIN_PASSWORD) != null) {
-                pswd = System.getProperty(LOGIN_PASSWORD).toCharArray();
+                incomingPassword = System.getProperty(LOGIN_PASSWORD).toCharArray();
             }
             doSet = true;
         }
 
         if (doSet) {
```

Comment:
may be a little easier to understand as if(incomingUsername == null || incomingPassword ==null) { and do away with "doSet" and assignments.